stdlib/public/TensorFlow/TensorGroup.swift

```diff
@@ -67,7 +67,8 @@ public extension OutputTensorGroup {
 ///
 /// TODO: Add a derived conformance to TensorGroup so that users don't have
 /// to write the conformance themselves.
-public protocol TensorGroup : InputTensorGroup & OutputTensorGroup {}
+
+public typealias TensorGroup = InputTensorGroup & OutputTensorGroup
```

Comment:
@rxwei You alluded that you had better names for these. Would you like to discuss that here?

It was not my intent, however, to make InputTensorGroup and OutputTensorGroup public other than my additional Array conformance.

It seems like I misunderstood the significance of the "///" comments. I could make them private by converting them all to "//", but then "TensorGroup" (The public facing API) would have no public documentation.

Reply:
I was suggesting that there could be better names. And there better has to be because "input" and "output" are not very meaningful in Swift. In Swift, function inputs are called "arguments" and outputs are called "results". That, however, doesn't make "argument" and "result" better names because "tensor operation" is a non-meaningful word in Swift either.

To give it an acceptable API name, i think the heart of the matter is whether the capacity of a tensor group is variable-sized or not, and it has nothing to do with input/outputs from a user perspective. In that case, names `TensorGroup` and `DynamicTensorGroup` bring it a bit closer. Nonetheless, it'd be better to discuss this over a meeting.

> It was not my intent, however, to make InputTensorGroup and OutputTensorGroup public other than my additional Array conformance.

We have to expose `TensorGroup` to the user to enable automatic packing/unpacking. This means `TensorGroup` is just as visible (and supported) as `Tensor`. And we can't make a public protocol refine non-public protocols - this is why we have many protocol names and method names with a prepended underscore in both Swift stdlib and the TF module, so that code completion won't show these to the user.

> It seems like I misunderstood the significance of the "///" comments. I could make them private by converting them all to "//", but then "TensorGroup" (The public facing API) would have no public documentation.

Triple-slash comments are doc comments that will be rendered in API docs. Double-slash comments won't and are for implementers. The original doc comment was okay, but the doc you added was more implementation-focused so they are better suited for double-slash comments.

Reply:
Let's discuss more over video today.